src/hotspot/share/runtime/handshake.cpp

```diff
@@ -667,8 +680,15 @@ class SuspendThreadHandshake : public HandshakeClosure {
 };
 
 bool HandshakeState::suspend() {
+  JavaThread* self = JavaThread::current();
   SuspendThreadHandshake st;
   Handshake::execute(&st, _handshakee);
+  if (_handshakee == self) {
+    // If target is the current thread we need to call this to do the
+    // actual suspend since Handshake::execute() above only installed
+    // the asynchronous handshake.
+    SafepointMechanism::process_if_requested(self);
+  }
```

Comment:
Is this an optimization? Or can the thread escape the suspend request?

Reply:
For self-suspend requests coming from Java we will suspend when going back to Java. In the case of JVMTI self-suspend requests I think Robbin found some JVMTI tests that were expecting to not return back to native without suspending.

Reply:
We should document the order of events why this code is needed in a future RFE, because this comment doesn't really explain it to me and knowing where to put all these process_if_requested() and process() calls is far from obvious.  I believe you and Robbin, but there are too many mysteries.

Reply:
I commented the line and run all the serviceability/ tests to look for issues. The only test that failed was SuspendWithCurrentThread.java due to the recent 8264663 change (https://github.com/openjdk/jdk/commit/b5c6351315eeec64fe9c06fc7c8719efa08dc8a6). The patch aimed to verify that the self-suspending thread actually suspended and didn't exited, but the check is too restrictive because it doesn't consider that the JT will still suspend when going back to Java. That check can be moved to Java to verify the suspender didn't return from suspendTestedThreads().
Maybe I should do that in another RFE along with removing this code?

Reply:
Yes, that would be good to clear up in a future RFE.